client/galaxy/scripts/mvc/tool/tool-form-composite.js

```diff
@@ -319,6 +319,19 @@ define([ 'utils/utils', 'utils/deferred', 'mvc/ui/ui-misc', 'mvc/form/form-view'
             }
         },
 
+        /** Refresh the history after job submission while form is shown */
+        _refreshHistory: function() {
+            var self = this;
+            var history = parent.Galaxy && parent.Galaxy.currHistoryPanel && parent.Galaxy.currHistoryPanel;
+            this._refresh_history && clearTimeout( this._refresh_history );
+            if ( history ) {
+                history.refreshContents();
+                if ( history.model.numOfUnfinishedShownContents() === 0 ) {
+                    this._refresh_history = setTimeout( function() { self._refreshHistory() }, 2000 );
+                }
+            }
+        },
+
         /** Build remaining steps */
         _execute: function() {
```

Comment:
@dannon You're right - if the jobs were intermittent while the workflow was running, the history would stop polling.

I'm obviously no expert on the WorkflowInvocation stuff, but I'd think (ideally) you'd want to poll there (a WorkflowInvocation api) first, getting some sort of 'running' in WorkflowInvocation.state (from the workflow UI) before you started polling the history. But there doesn't seem to be a 'running' or in-progress state listed there.

Of course, most of these problems go away if we can ever get a backend that can do websockets well - so that would be an option, too.